### ISSUE TITLE
Disable Btrfs async discard by default?

### ISSUE THREAD

**Issue by HanabishiRecca** [zen-kernel/zen-kernel#337]
As you may know, since the kernel version 6.2 `discard=async` mount option is set by default for Btrfs (63a7cb13071842966c1ce931edacbc23573aada5). And in 6.3 the default discard `iops_limit` has been raised to 1000.

That decision was pushed by Facebook devs, based on their server workloads. But having it enabled by default on desktop does not seem really beneficial in comparison with periodic TRIMs, and in some cases may be even decremental. Read [this discussion](https://lore.kernel.org/linux-btrfs/CAEzrpqeOAiYCeHCuU2O8Hg5=xMwW_Suw1sXZtQ=f0f0WWHe9aw@mail.gmail.com/t/#u) for details.

Some highlights:

> We got the defaults based on our testing with our workloads inside of
FB.  Clearly this isn't representative of a normal desktop usage, but
we've also got a lot of workloads so figured if it made the whole
fleet happy it would probably be fine everywhere.

> FYI, discard performance differs a lot between different SSDs.
It used to be pretty horrible for most devices early on, and then a
certain hyperscaler started requiring decent performance for enterprise
drives, so many of them are good now.  A lot less so for the typical
consumer drive, especially at the lower end of the spectrum.

> For discard things have gotten a lot better in the last years, but for
many older devices performance can be outright horrible.

> For SATA SSDs the fact that classic TRIM isn't a queued command adds
insult to injury as it always means draining the queue first and not
issuing any I/O until the TRIM command is done.  There is a FPDMA
version not, but I don't think it ws all that widely implemented before
SATA SSDs fell out of favour.

> Will power usage be taken into consideration? I only noticed this regression
myself when my laptop started to draw a constant extra ~10W from the constant
drive activity, so I imagine other laptop users would also prefer a default
which avoids this if possible. If "relatively quickly" means anything more
than a few seconds I could see that causing rather significant battery life
reductions.

To summarize:

* Old/cheap consumer-grade SSDs can suffer from constant disards because of poor TRIM implementation. From latency spikes to complete I/O blocking.
* Laptops can issue power draw increase because of constant drive activity.

So I have a proposal to disable Btrfs `discard=async` default for Zen kernel, as it primarily targets desktop users.

Unfortunately, this thing does not have a config tunable, so it can only be disabled by patching the source.

**Comment by HanabishiRecca** [comment]
> so it can only be disabled by patching the source.

Basically

```patch
diff --git a/fs/btrfs/disk-io.c b/fs/btrfs/disk-io.c
index b79781df7..5c6326a54 100644
--- a/fs/btrfs/disk-io.c
+++ b/fs/btrfs/disk-io.c
@@ -3492,23 +3492,6 @@ int __cold open_ctree(struct super_block *sb, struct btrfs_fs_devices *fs_device
 		btrfs_set_and_info(fs_info, SSD, "enabling ssd optimizations");
 	}
 
-	/*
-	 * For devices supporting discard turn on discard=async automatically,
-	 * unless it's already set or disabled. This could be turned off by
-	 * nodiscard for the same mount.
-	 *
-	 * The zoned mode piggy backs on the discard functionality for
-	 * resetting a zone. There is no reason to delay the zone reset as it is
-	 * fast enough. So, do not enable async discard for zoned mode.
-	 */
-	if (!(btrfs_test_opt(fs_info, DISCARD_SYNC) ||
-	      btrfs_test_opt(fs_info, DISCARD_ASYNC) ||
-	      btrfs_test_opt(fs_info, NODISCARD)) &&
-	    fs_info->fs_devices->discardable &&
-	    !btrfs_is_zoned(fs_info)) {
-		btrfs_set_and_info(fs_info, DISCARD_ASYNC,
-				   "auto enabling async discard");
-	}
 
 #ifdef CONFIG_BTRFS_FS_CHECK_INTEGRITY
 	if (btrfs_test_opt(fs_info, CHECK_INTEGRITY)) {
```

works.

**Comment by heftig** [comment]
I'm a happy user of async discard and don't want to optimize the defaults to favor old hardware.

**Comment by heftig** [comment]
In addition, I think changing the default to `nodiscard` is actively harmful here, as users might not have periodic trim configured. Anyone who sets up periodic trim can also easily add the `nodiscard` parameter.

**Comment by HanabishiRecca** [comment]
> I'm a happy user of async discard and don't want to optimize the defaults to favor old hardware.

Fair, but

1. You won't suffer from disabling it as much, but old/cheap hardware users can suffer a lot.
2. It is really benefitial only for extensive server workloads, where periodic TRIMs are suboptimal.
3. Users can always enable it via mount options.

Also having such things opt-in is better IMO.
